### PR TITLE
[SPARK-14215] [SQL] [PYSPARK] Support chained Python UDFs

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

This PR brings the support for chained Python UDFs, for example

``` sql
select udf1(udf2(a))
select udf1(udf2(a) + 3)
select udf1(udf2(a) + udf3(b)) 
```

Also directly chained unary Python UDFs are put in single batch of Python UDFs, others may require multiple batches.

For example,

``` python
>>> sqlContext.sql("select double(double(1))").explain()
== Physical Plan ==
WholeStageCodegen
:  +- Project [pythonUDF#10 AS double(double(1))#9]
:     +- INPUT
+- !BatchPythonEvaluation double(double(1)), [pythonUDF#10]
   +- Scan OneRowRelation[]
>>> sqlContext.sql("select double(double(1) + double(2))").explain()
== Physical Plan ==
WholeStageCodegen
:  +- Project [pythonUDF#19 AS double((double(1) + double(2)))#16]
:     +- INPUT
+- !BatchPythonEvaluation double((pythonUDF#17 + pythonUDF#18)), [pythonUDF#17,pythonUDF#18,pythonUDF#19]
   +- !BatchPythonEvaluation double(2), [pythonUDF#17,pythonUDF#18]
      +- !BatchPythonEvaluation double(1), [pythonUDF#17]
         +- Scan OneRowRelation[]
```

TODO: will support multiple unrelated Python UDFs in one batch (another PR).
## How was this patch tested?

Added new unit tests for chained UDFs.

### CODE REVIEW

**Comment by davies** [comment]
cc @marmbrus @rxin

**Comment by davies** [comment]
cc @cloud-fan

**Comment by hvanhovell** [comment]
@davies I think the JIRA number should be SPARK-14215: https://issues.apache.org/jira/browse/SPARK-14215

**Comment by davies** [comment]
@hvanhovell Corrected, thanks!

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -1648,6 +1648,14 @@ def sort_array(col, asc=True):
 
 # ---------------------------- User Defined Function ----------------------------------
 
+def _wrap_function(sc, func, returnType):
```
Comment (cloud-fan) [inline python/pyspark/sql/functions.py]: we already have a `_wrap_function` in rdd.py

Comment (cloud-fan) [inline python/pyspark/sql/functions.py]: what's the point of creating a new `_wrap_function` here? To decrease the size of serialized python function?

Comment (cloud-fan) [inline python/pyspark/sql/functions.py]: oh i see, we wanna chain the functions at python worker side.

**Comment by cloud-fan** [comment]
Overall LGTM

File: python/pyspark/worker.py [hunk 2]
```diff
@@ -95,10 +107,23 @@ def main(infile, outfile):
...
 
         _accumulatorRegistry.clear()
-        command = pickleSer._read_with_length(infile)
-        if isinstance(command, Broadcast):
-            command = pickleSer.loads(command.value)
-        func, profiler, deserializer, serializer = command
+        row_based = read_int(infile)
+        num_commands = read_int(infile)
+        if row_based:
+            profiler = None  # profiling is not supported for UDF
```
Comment (viirya) [inline python/pyspark/worker.py]: `profiler` seems need to be defined before this if block. The codes refer `profiler` later out of this block.

Comment (davies) [inline python/pyspark/worker.py]: The other branch also have `profiler`, so I think it's fine.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala [hunk 3]
```diff
@@ -29,13 +30,31 @@ import org.apache.spark.sql.catalyst.rules.Rule
...
+      case udf: PythonUDF if canEvaluate(udf) => udf
+    }
+  }
+
   def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
     // Skip EvaluatePython nodes.
     case plan: EvaluatePython => plan
 
     case plan: LogicalPlan if plan.resolved =>
       // Extract any PythonUDFs from the current operator.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala]: We should update the comments the explain our new strategy of extracting and evaluating python udfs.

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala]: Updated.

**Comment by davies** [comment]
Merging this into master (the last commit only added comments).

**Comment by gatorsmile** [comment]
@davies It sounds you used a wrong JIRA number.

**Comment by davies** [comment]
@gatorsmile Corrected in PR, but the notification email is not updated.

**Pull request closed without merging**